### ISSUE TITLE
Camel routes and source are not displayed for Camel 2.17.3

### ISSUE THREAD

**Issue by davidfra** [hawtio/hawtio#2248]
Hi,
I recently updated my application from Camel version 2.17.0 to version 2.17.3.
I could not go for 2.18.0 because it would require some more refactoring.

But now, in hawtio I cannot display the route diagrams and source view any more.
It just displays an empty page. This also makes debugging of routes impossible.

I tried hawtio 1.4.65 and 1.4.67, but they both show the same result.

**Comment by davsclaus** [comment]
Can you try with 2.17.1 and 2.17.2 and see if any of them works for you?

**Comment by davidfra** [comment]
I did a little testing to try to narrow it down. It seems the Camel version is not the problem after all. 
I tried it with Camel 2.17.0 again and got the same Error. 
As we are constantly developing our application it is hard to pin down the exact cause.

What I found is that calling `dumpRoutesAsXml( )` on the context MBean throws a `ClassCastException`.
It is the same when calling `dumpRouteAsXml( )` on some of the routes. For some routes I get the correct XML dump.

```
Caused by: java.lang.ClassCastException: java.util.HashSet cannot be cast to java.lang.String
	at com.sun.xml.internal.bind.v2.runtime.XMLSerializer.attWildcardAsAttributes(XMLSerializer.java:764)
	at com.sun.xml.internal.bind.v2.runtime.ClassBeanInfoImpl.serializeAttributes(ClassBeanInfoImpl.java:380)
	at com.sun.xml.internal.bind.v2.runtime.XMLSerializer.childAsSoleContent(XMLSerializer.java:576)
	at com.sun.xml.internal.bind.v2.runtime.ClassBeanInfoImpl.serializeRoot(ClassBeanInfoImpl.java:326)
	at com.sun.xml.internal.bind.v2.runtime.property.ArrayReferenceNodeProperty.serializeListBody(ArrayReferenceNodeProperty.java:103)
	at com.sun.xml.internal.bind.v2.runtime.property.ArrayERProperty.serializeBody(ArrayERProperty.java:144)
	at com.sun.xml.internal.bind.v2.runtime.ClassBeanInfoImpl.serializeBody(ClassBeanInfoImpl.java:345)
	at com.sun.xml.internal.bind.v2.runtime.XMLSerializer.childAsSoleContent(XMLSerializer.java:578)
	at com.sun.xml.internal.bind.v2.runtime.ClassBeanInfoImpl.serializeRoot(ClassBeanInfoImpl.java:326)
	at com.sun.xml.internal.bind.v2.runtime.XMLSerializer.childAsRoot(XMLSerializer.java:479)
	at com.sun.xml.internal.bind.v2.runtime.MarshallerImpl.write(MarshallerImpl.java:308)
	at com.sun.xml.internal.bind.v2.runtime.MarshallerImpl.marshal(MarshallerImpl.java:236)
	at javax.xml.bind.helpers.AbstractMarshallerImpl.marshal(AbstractMarshallerImpl.java:116)
	at org.apache.camel.model.ModelHelper.dumpModelAsXml(ModelHelper.java:59)
	at org.apache.camel.management.mbean.ManagedCamelContext.dumpRoutesAsXml(ManagedCamelContext.java:411)
	at org.apache.camel.management.mbean.ManagedCamelContext.dumpRoutesAsXml(ManagedCamelContext.java:398)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
       ......
```

The Camel context is setup inside an OSGi context which is embedded in a webapp. We use the Java DSL to build the routes.
I couldn't yet find out, what is different about the routes for that the XML dump is not working.

**Comment by davidfra** [comment]
I found the reason.

I set an attribute on some route definitions using the method `attribute(QName name, Object value)`.
Where `value` is a `Set<String>`.

This seems to be the reason for the Exception when calling `dumpRoutesAsXml( )`.
When I remove the attributes, the dump works and hawtio displays the source and the route diagram.

So it seems to me this is more a Camel bug than a hawtio bug.

**Comment by davsclaus** [comment]
What are you doing and why are you doing that?

**Comment by davidfra** [comment]
Well I added a `RoutePolicyFactory` to the Camel context that must ensure that the user that calls the route has the needed permissions.

Therefore I set a custom attribute on the `RouteDefinition` that contains the needed permissions.
For example:

```
from("direct:myroute")
	.attribute(new QName("namespaceURI", "needed-permissions"), Collections.singleton("manage-accounts"))
	.to("something");
```

Then my custom `RoutePolicy` checks if the user has the permission **manage-accounts** every time the route **direct:myroute** is called. I use a Set<String> of needed permissions so I have the possibility to define more than one.

This works great, but throws an Exception when calling `dumpRouteAsXml()`.
So I consider using a comma separated String instead.

But I am asking myself why the method `RouteDefinition.attribute(QName name, Object value)` accepts an Object as value if it only works with Strings (or at least no with any possible Object).

Or is there a better way to add custom attributes to a RouteDefinition that can be checked by the `RoutePolicy`?
